docs/how-to/push-update-on-mainnet.md

```diff
@@ -0,0 +1,120 @@
+do (`0` states for first key imported).
+* Parameters 2-5 are block version data parameters -- leave them as
+they are provided in the cli example.
+* `csl-daedalus:1` is software version description, you should substitute `1` (version) with the integer provided along with installers (see *Prerequisites* section, *4.* item)
+* End of command: list of triples -- update installers. Provide the path to installers
+without any slashes (installers should be in the same dir as auxx is
+launched in). First tuple element is platform, second is path, `none` stands for binary diff (this feature is not used for now).
+
+That's it. Successfull command output looks like this:
+
+```
+[smart-wallet:DEBUG] [2017-09-21 15:13:24 MSK] Proposing update...
+Read file installer062win64.exe succesfuly, its hash: 01abf1c8b881c2f8ea4d1349a700f29d4088e68dc04b6bf4680ea7e14638373e
+Read file installer062macos64.pkg succesfuly, its hash: 3bc1084841fb99fff03ef92bc35eef9a80a20aeca688dfc1b50a4aa6dd6f7c73
+[smart-wallet:INFO] [2017-09-21 15:13:25 MSK] Announcing proposal with id 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6
+Update proposal submitted, upId: 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6
+```
+
+Note:
+
+```
+Read file installer062win64.exe succesfuly, its hash: 01abf1c8b881c2f8ea4d1349a700f29d4088e68dc04b6bf4680ea7e14638373e
+Read file installer062macos64.pkg succesfuly, its hash: 3bc1084841fb99fff03ef92bc35eef9a80a20aeca688dfc1b50a4aa6dd6f7c73
+```
+
+These hashes should be used as names when you will upload installers on S3 buckets.
+
+Note: 
+```
+Update proposal submitted, upId: 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6
+```
+
+Value `upId` is used on one of next steps, when we'll vote for update.
+
+Uploading update files
+======================
+
+Upload installers to S3 bucket:
+* `update-system-testing` if update is being performed on staging
+* `update-system` if update is being performed on live mainnet cluster
+
+Voting for proposal
+===================
+
+```
+stack exec cardano-auxx -- --system-start 0 --log-config
+  scripts/log-templates/log-config-qa.yaml
+  --logs-prefix "logs/auxx-update-1.0.1" --db-path auxx-update-1.0.1
+  --peer <relay dns name>:3000 cmd --commands "vote 1 y 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6,vote 2 y 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6,vote 3 y 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6,vote 4 y 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6"
+```
+
+In `vote N y 4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6`:
+
+* `N` should be index of key (from `listadr`)
+* `y` is for yes
+* `4c827d6fe03d4c3646ebbbc28d4e09c57690e1dcba54b9adc0050d3f76734cf6` is update proposal id or `upId`
+
+Successfull output ends in "submitted a vote". Votes will be sent to the network, software update will apply soon.
```

Comment:
1. What mechanism guarantees that proposed updates which haven't yet been approved by a majority can't possibly result in a software update?
2. How do we check whether a node has applied an update yet?
3. If a node fails to automatically apply an update (or fails during the update process) is there a recommended manual update procedure?
4. Can we check what ratio of the network is running the current set of updates?
5. Suppose while a node is offline multiple updates are proposed and voted in. When the node comes back online, does it apply the updates sequentially in the order they were ratified, a random one, only the latest one, or something else? Does this strategy ensure the node's final state is valid?

Reply:
> What mechanism guarantees that proposed updates which haven't yet been approved by a majority can't possibly result in a software update?

There's no ironclad guarantee for this, of course, but we only download confirmed proposals and we only call `addConfirmedProposal` on a proposal if (according to our calculations) it has been approved by stake majority.

Reply:
> If a node fails to automatically apply an update (or fails during the update process) is there a recommended manual update procedure?

Just run the newest installer. You might also need to delete `%APPDATA%\Daedalus` and restore your wallet afterwards.

> How do we audit the update/vote information written to the blockchain?

It looks like auxx doesn't have a command like “show all available proposals”, but it should be easy to add. Otherwise, we can just look at another node's logs – when it prints information about received blocks, it should also show the votes/proposals contained in those blocks.

Reply:
> **2.** How do we check whether a node has applied an update yet?

A particular node that you have access to?
* It will restart itself (which should be visible in the logs)
* After a restart it will print its version (e.g. `Application: cardano-sl:0`), which too will be visible in the logs

> **4.** Can we check what ratio of the network is running the current set of updates?

AFAIK no. We can try looking at users' logs from the report server, but I'm not sure that we have enough info about each user to avoid double-counting some of them.

Reply:
> **5.** When the node comes back online, does it apply the updates sequentially in the order they were ratified, select a random one, select only the latest one, or something else? Does this strategy ensure the node's final state is valid?

The worker which applies the last downloaded update runs concurrently with the worker which downloads updates, so it might end up choosing a random one. E.g. if there are two confirmed updates, then both of these options are possible:

* download v1, restart & apply v1, download v2, restart & apply v2
* download v1, download v2, restart & apply v2

One thing that we can't do is apply v1 after v2, so the state of the node will be sound if all of {v0 → v1, v1 → v2, v0→v2} are valid. If you don't test v0→v2, then you would be relying on luck.

Reply:
> One thing that we can't do is apply v1 after v2

Is this scenario prevented by code? I realize it would be bad, but I'm unclear whether we need to do anything special to avoid it.

Reply:
> If you don't test v0→v2, then you would be relying on luck.

Can we seriously consider enforcing consecutive version updates in the code to prevent this? As the number of updates grows the number of combinations for possible update paths grows rapidly.

Reply:
> Is [applying v2 after v1] prevented by code?

Yep. We know our software version and don't apply anything that isn't newer.

Reply:
> don't apply anything that isn't newer.

That's not my concern. My concern is skipping versions.

Reply:
I gather from https://github.com/input-output-hk/cardano-sl/pull/1697/files#r142403318 that proposals which skip versions are prevented.

Reply:
Ah, but even if all ratified proposals have consecutive versions, updates may be applied by skipping versions. My earlier concern still stands.

>> If you don't test v0→v2, then you would be relying on luck.

> Can we seriously consider enforcing consecutive version updates in the code to prevent this? As the number of updates grows the number of combinations for possible update paths grows rapidly.

Reply:
We always download the latest update, consecutive updates are not supposed to be supported for now. We distribute installers (for now, not binary diffs) so it shouldn't be a problem.

Consecutive updates are really tedious for those who need to update for a lot of versions since every update is basically "kill node, run executable, start node over".

Reply:
@volhovM, @neongreen, sorry, but aren't these two statements in a more-or-less direct contradiction?

statement 1:
> The worker which applies the last downloaded update runs concurrently with the worker which downloads updates, so it might end up choosing a random one. E.g. if there are two confirmed updates, then both of these options are possible:
>
>    download v1, restart & apply v1, download v2, restart & apply v2
>    download v1, download v2, restart & apply v2

statement 2:

> We always download the latest update, consecutive updates are not supposed to be supported for now.

?

Reply:
https://github.com/input-output-hk/cardano-sl/blob/master/lib/src/Pos/Update/Worker.hs#L67
We download the newest one.

Reply:
@denisshevchenko these questions about the update system are really useful, I suggest we incorporate them into the cardano docs.